programs/mango-v4/src/instructions/token_force_close_borrows_with_token.rs

```diff
@@ -0,0 +1,207 @@
+
+    require_keys_neq!(ctx.accounts.liqor.key(), ctx.accounts.liqee.key());
+    let mut liqor = ctx.accounts.liqor.load_full_mut()?;
+    // account constraint #1
+    require!(
+        liqor
+            .fixed
+            .is_owner_or_delegate(ctx.accounts.liqor_owner.key()),
+        MangoError::SomeError
+    );
+    require_msg_typed!(
+        !liqor.fixed.being_liquidated(),
+        MangoError::BeingLiquidated,
+        "liqor account"
+    );
+
+    let mut liqee = ctx.accounts.liqee.load_full_mut()?;
+    let mut liqee_health_cache = new_health_cache(&liqee.borrow(), &account_retriever)
+        .context("create liqee health cache")?;
+    let liqee_liq_end_health = liqee_health_cache.health(HealthType::LiquidationEnd);
+
+    //
+    // Transfer liab_token from liqor to liqee to close the borrows.
+    // Transfer corresponding amount of asset_token from liqee to liqor.
+    //
+    let now_ts = Clock::get()?.unix_timestamp.try_into().unwrap();
+    {
+        let account_retriever: &mut ScanningAccountRetriever = &mut account_retriever;
+        let liab_token_index = liab_token_index;
+        let asset_token_index = asset_token_index;
+        let liqor: &mut MangoAccountRefMut = &mut liqor.borrow_mut();
+        let liqor_key = ctx.accounts.liqor.key();
+        let liqee: &mut MangoAccountRefMut = &mut liqee.borrow_mut();
+        let liqee_key = ctx.accounts.liqee.key();
+
+        let (asset_bank, asset_oracle_price, opt_liab_bank_and_price) =
+            account_retriever.banks_mut_and_oracles(asset_token_index, liab_token_index)?;
+        let (liab_bank, liab_oracle_price) = opt_liab_bank_and_price.unwrap();
+        // account constraint #2
+        require!(liab_bank.is_force_close(), MangoError::SomeError);
+
+        let (liqee_asset_position, liqee_asset_raw_index) =
+            liqee.token_position_and_raw_index(asset_token_index)?;
+        let liqee_asset_native = liqee_asset_position.native(asset_bank);
+        // TODO: should we enforce that asset position is positive? or just incur borrows if needed, I think borrowing is fine
+
+        let (liqee_liab_position, liqee_liab_raw_index) =
+            liqee.token_position_and_raw_index(liab_token_index)?;
+        let liqee_liab_native = liqee_liab_position.native(liab_bank);
+        require!(liqee_liab_native.is_negative(), MangoError::SomeError);
+
+        // The amount of liab native tokens we will transfer
+        let liab_transfer = min(-liqee_liab_native, max_liab_transfer);
+
+        // The amount of asset native tokens we will give up for them
+        let fee_factor = I80F48::ONE + liab_bank.liquidation_fee;
+        let liab_oracle_price_adjusted = liab_oracle_price * fee_factor;
+        let asset_transfer = liab_transfer * liab_oracle_price_adjusted / asset_oracle_price;
+
+        // Apply the balance changes to the liqor and liqee accounts
+        let liqee_liab_position = liqee.token_position_mut_by_raw_index(liqee_liab_raw_index);
```

Comment:
why retrieve it again? we still have it in-scope! could we get the mut version from the start?